Timetables guide! (#281)
astronomer/airflow-guides: guides/scheduling-in-airflow.md

date: 2021-10-13T00:00:00.000Z
slug: "scheduling-in-airflow"
heroImagePath: null
tags: ["DAGs"]
---

## Overview

One of the most fundamental features of Apache Airflow is the ability to schedule jobs. Historically, Airflow users could schedule their DAGs by specifying a `schedule_interval` with a cron expression, timedelta, or using a preset Airflow schedule. Then with the release of Airflow 2.2 in October 2021, scheduling flexibility was massively increased with the introduction of timetables. Timetables allow users to create their own custom schedules using Python, eliminating gaps in scheduling capability previously caused by limitations of cron. Now no matter what use case your DAG is serving, it can be scheduled in Airflow.

In this guide, we'll walk through Airflow scheduling concepts and the different ways you can schedule a DAG with a focus on new features released in Airflow 2.2.  

> Note: All code in this guide can be found in [this repo](https://github.com/astronomer/airflow-scheduling-tutorial).

## Scheduling Concepts

There are a couple of concepts and terms in Airflow that are important to understand related to scheduling:

> - **Data Interval**: The period of data that a task should operate on. This concept is made up of two parameters, both of which are `datetime` objects:

Author: Re-orged this section to have subsections for general terms vs parameters that can be defined, and expanded on some of the definitions. Hopefully this makes it clearer.

![5 Minute Example DAG](https://assets2.astronomer.io/main/guides/scheduling-in-airflow/5_min_example.png)

If we looks at the next run in the UI, which represents the *next* logical date, it shows `2021-10-08 19:17:36`. This is 5 minutes after the previous logical date, and the same as the data interval end of the last DAG Run since there are no "holes" in the schedule. The data interval of the next DAG run is also shown, and Run After, which is the date and time the DAG Run will be scheduled, is the same as the end of the data interval `2021-10-08 19:22:36`.

![5 Minute Next Run](https://assets2.astronomer.io/main/guides/scheduling-in-airflow/5_min_next_run.png)

In the sections below, we'll walk through how to use the `schedule_interval` or timetables to schedule your DAG.

## Schedule Interval

For pipelines with basic schedules, you can define a `schedule_interval` in your DAG. Prior to Airflow 2.2, this was the *only* mechanism for defining a DAG's schedule. 

There are multiple ways you can define the `schedule_interval`:

> - **Cron expression:** `schedule_interval = '0 0 * * *'`

Reviewer: ```suggestion
### Cron expression`
```
Let's make these into `###` headers instead of bullet points. The examples following each method suffice for showing the formatting.

    
    If you need help creating the correct cron expression, [crontab guru](https://crontab.guru/) is a great resource.
    
- **Cron presets:** `schedule_interval = '@daily'`
    
    Airflow can utilize cron presets for common, basic schedules. 
    
    For example, `schedule_interval='@hourly'` will schedule the DAG to run at the beginning of every hour. For the full list of presets, check out the [Airflow documentation](https://airflow.apache.org/docs/apache-airflow/1.10.1/scheduler.html#dag-runs).
    
    Note: If your DAG does not need to run on a schedule and will only be triggered manually or externally triggered by another process, you can set `schedule_interval=None`.
    
- **Timedelta:** `schedule_interval=timedelta(days=1)`
    
    If you want to schedule your DAG on a particular cadence (hourly, every 5 minutes, etc.) rather than at a specific time, you can pass a `timedelta` object to the schedule interval. For example, `schedule_interval=timedelta(minutes=30)` will run the DAG every thirty minutes.
    

> > Note: Do not make your DAG's schedule dynamic (e.g. `datetime.now()`)! This will cause an error in the scheduler.

Reviewer: ```suggestion
> Note: Do not make your DAG's schedule dynamic (e.g. `datetime.now()`)! This will cause an error in the Scheduler.
```
```suggestion
> Note: Do not make your DAG's schedule dynamic (e.g. `datetime.now()`)! This will cause an error in the scheduler.
```

## Timetables

> The scheduling options detailed in the section above cover many common use cases, but they do have some limitations. For example, it is difficult or impossible to implement situations like:

Reviewer: I would introduce Timetables first, then discuss the limitations they eliminate. Either that, or add the limitations as an `###` header to the end of the previous section - up to you.

- Schedule a DAG at different times on different days, like 2pm on Thursdays and 4pm on Saturdays.
- Schedule a DAG daily except for holidays.
- Schedule a DAG at multiple times daily with uneven intervals (e.g. 1pm and 4:30pm).

> [Timetables](https://airflow.apache.org/docs/apache-airflow/stable/concepts/timetable.html), introduced in Airflow 2.2, address these limitations by allowing users to define their own schedules in Python code. All DAG schedules are determined by their internal timetable. You can still define a `schedule_interval`, which Airflow converts to a timetable behind the scenes. Or, if you have a use case like one of the ones listed above where a cron expression or timedelta is not sufficient, you can define your own timetable.

Reviewer: ```suggestion
[Timetables](https://airflow.apache.org/docs/apache-airflow/stable/concepts/timetable.html), introduced in Airflow 2.2, address these limitations by allowing users to define their own schedules in Python code. All DAG schedules are determined by their internal Timetable. 

Going forward, Timetables will become the primary method for scheduling in Airflow. You can still define a `schedule_interval`, but Airflow converts this to a Timetable behind the scenes. And if a cron expression or timedelta is not sufficient for your use case, you can define your own Timetable.
```

> Custom timetables can be registered as part of an Airflow plugin. They must be a subclass of `Timetable`, and should contain the following methods, both of which need to return a `DataInterval` with start and end:

Reviewer: ```suggestion
Custom timetables can be registered as part of an Airflow plugin. They must be a subclass of `Timetable`, and they should contain the following methods, both of which return a `DataInterval` with a start and an end:
```

> - `next_dagrun_info`: Method that returns the data interval for the DAG's regular schedule.
> - `infer_manual_data_interval`: Method that returns the data interval when the DAG is manually triggered.

Reviewer: ```suggestion
- `next_dagrun_info`: Returns the data interval for the DAG's regular schedule
- `infer_manual_data_interval`: Returns the data interval when the DAG is manually triggered
```

Below we'll show an example of implementing these methods in a custom timetable.

### Example Custom Timetable

> For this implementation, let's take the example of running a DAG multiple times a day but with uneven intervals: we want to run our DAG at 6:00 and 16:30. Because this schedule has run times with differing hours *and* minutes, it can't be represented by a single cron expression. But we can easily implement this schedule with a custom timetable!

Reviewer: ```suggestion
For this implementation, let's run our DAG at 6:00 and 16:30. Because this schedule has run times with differing hours *and* minutes, it can't be represented by a single cron expression. But we can easily implement this schedule with a custom timetable!
```

> To start, we need to define our `next_dagrun_info` and `infer_manual_data_interval` methods. Before diving into the code, it's helpful to think through what the data intervals will be for the schedule we want. Remember that the time the DAG runs (`run_after`) should be the *end* of the data interval since our interval has no holes. So in this case, for a DAG that we want to run at 6:00 and 16:30, we will have two different alternating intervals:

Reviewer: ```suggestion
To start, we need to define the `next_dagrun_info` and `infer_manual_data_interval` methods. Before diving into the code, it's helpful to think through what the data intervals will be for the schedule we want. Remember that the time the DAG runs (`run_after`) should be the *end* of the data interval since our interval has no holes. So in this case, for a DAG that we want to run at 6:00 and 16:30, we have two different alternating intervals:
```

> - Run at 6:00: interval is from 16:30 *the previous day* to 6:00 the current day

Reviewer: ```suggestion
- Run at 6:00: Data interval is from 16:30 *the previous day* to 6:00 the current day
```

> - Run at 16:30: interval is from 6:00 to 16:30 the current day

Reviewer: ```suggestion
- Run at 16:30: Data interval is from 6:00 to 16:30 the current day
```

        else: # If previous period started at 14:30, next period will start at 6:00 next day and end at 14:30
            next_start = (last_start+delta).set(hour=6, minute=0).replace(tzinfo=UTC)
            next_end = (last_start+delta).replace(tzinfo=UTC)
    else:  # This is the first ever run on the regular schedule. First data interval will always start at 6:00 and end at 16:30
        next_start = restriction.earliest
        if next_start is None:  # No start_date. Don't schedule.
            return None
        if not restriction.catchup: # If the DAG has catchup=False, today is the earliest to consider.
            next_start = max(next_start, DateTime.combine(Date.today(), Time.min).replace(tzinfo=UTC))
        next_start = next_start.set(hour=6, minute=0).replace(tzinfo=UTC)
        next_end = next_start.set(hour=16, minute=30).replace(tzinfo=UTC)
    if restriction.latest is not None and next_start > restriction.latest:
        return None  # Over the DAG's scheduled end; don't schedule.
    return DagRunInfo.interval(start=next_start, end=next_end)
```

Walking through the logic, this code is equivalent to:

- If there was a previous run for the DAG:

>     - If the last DAG Run started at 6:00, then the next DAG run should start at 16:30 and end at 6:00 the next day.

Reviewer: ```suggestion
    - If the previous DAG Run started at 6:00, then the next DAG run should start at 16:30 and end at 6:00 the next day.
```

>     - If the last DAG run started at 14:30, then the DAG run should start at 6:00 the next day and end at 16:30 the next day.

Reviewer: ```suggestion
    - If the previous DAG run started at 16:30, then the DAG run should start at 6:00 the next day and end at 16:30 the next day.
```

    delta = timedelta(days=1)
    # If time is between 6:00 and 16:30, period ends at 6am and starts at 16:30 previous day
    if run_after >= run_after.set(hour=6, minute=0) and run_after <= run_after.set(hour=16, minute=30):
        start = (run_after-delta).set(hour=16, minute=30, second=0).replace(tzinfo=UTC)
        end = run_after.set(hour=6, minute=0, second=0).replace(tzinfo=UTC)
    # If time is after 16:30 but before midnight, period is between 6:00 and 16:30 the same day
    elif run_after >= run_after.set(hour=16, minute=30) and run_after.hour <= 23:
        start = run_after.set(hour=6, minute=0, second=0).replace(tzinfo=UTC)
        end = run_after.set(hour=16, minute=30, second=0).replace(tzinfo=UTC)
    # If time is after midnight but before 6:00, period is between 6:00 and 16:30 the previous day
    else:
        start = (run_after-delta).set(hour=6, minute=0).replace(tzinfo=UTC)
        end = (run_after-delta).set(hour=16, minute=30).replace(tzinfo=UTC)
    return DataInterval(start=start, end=end)
```

This method figures out what the most recent complete data interval is based on the current time. There are three scenarios:

> - The current time is between 6:00 and 16:30. In this case the data interval starts at 16:30 the previous day and ends at 6:00 the current day.
> - The current time is after 16:30 but before midnight. In this case the data interval is from 6:00 to 16:30 the current day.
> - The current time is after midnight but before 6:00. In this case the data interval is between 6:00 and 16:30 the previous day.

Reviewer: ```suggestion
- The current time is between 6:00 and 16:30: In this case, the data interval is from 16:30 the previous day to 6:00 the current day.
- The current time is after 16:30 but before midnight: In this case, the data interval is from 6:00 to 16:30 the current day.
- The current time is after midnight but before 6:00: In this case, the data interval is from 6:00 to 16:30 the previous day.
```

> Some of these periods are the same timeframe but on different days relevant to the time the DAG is triggered, requiring three sets of logic. When defining custom timetables, always keep in mind what the last complete data interval should be based on when the DAG should run.

Reviewer: ```suggestion
We need to account for time periods in the same timeframe (6:00 to 16:30) on different days than the day that the DAG is triggered, which requires three sets of logic. When defining custom timetables, always keep in mind what the last complete data interval should be based on when the DAG should run.
```

> Now we can take those two methods and combine them into a `Timetable` class which will create our Airflow plugin. The full custom timetable plugin is below:

Reviewer: ```suggestion
Now we can take those two methods and combine them into a `Timetable` class which will make up our Airflow plugin. The full custom timetable plugin is below:
```

                next_start = (last_start+delta).set(hour=6, minute=0).replace(tzinfo=UTC)
                next_end = (last_start+delta).replace(tzinfo=UTC)
        else:  # This is the first ever run on the regular schedule. First data interval will always start at 6:00 and end at 16:30
            next_start = restriction.earliest
            if next_start is None:  # No start_date. Don't schedule.
                return None
            if not restriction.catchup: # If the DAG has catchup=False, today is the earliest to consider.
                next_start = max(next_start, DateTime.combine(Date.today(), Time.min).replace(tzinfo=UTC))
            next_start = next_start.set(hour=6, minute=0).replace(tzinfo=UTC)
            next_end = next_start.set(hour=16, minute=30).replace(tzinfo=UTC)
        if restriction.latest is not None and next_start > restriction.latest:
            return None  # Over the DAG's scheduled end; don't schedule.
        return DagRunInfo.interval(start=next_start, end=next_end)

class UnevenIntervalsTimetablePlugin(AirflowPlugin):
    name = "uneven_intervals_timetable_plugin"
    timetables = [UnevenIntervalsTimetable]
```

> Note that because timetables are plugins, you will need to restart the Airflow scheduler and webserver after adding or updating them.

Reviewer: ```suggestion
Note that because timetables are plugins, you will need to restart the Airflow Scheduler and Webserver after adding or updating them.
```

```python
from uneven_intervals_timetable import UnevenIntervalsTimetable

with DAG(
    dag_id="example_timetable_dag",
    start_date=datetime(2021, 10, 9), 
    max_active_runs=1,
    timetable=UnevenIntervalsTimetable(),
    default_args={
        "retries": 1,
        "retry_delay": timedelta(minutes=3),
    },
    catchup=True
) as dag:
```

Looking at the Tree View in the UI, we can see that this DAG has run twice per day at 6:00 and 16:30 since the start date of 10/9/2021. 

![Timetable DAG Runs](https://assets2.astronomer.io/main/guides/scheduling-in-airflow/timetable_catchup_runs.png)

The next scheduled run is for the interval starting on 10/12/2021 at 16:30 and ending the following day at 6:00. This run will be triggered at the end of the data interval, so after 10/13/2021 6:00.

![Timetable Next Run](https://assets2.astronomer.io/main/guides/scheduling-in-airflow/timetable_next_run.png)

If we run the DAG manually after 16:30 but before midnight, we can see the data interval for the triggered run was between 6:00 and 16:30 that day as expected.

![Timetable Manual Run](https://assets2.astronomer.io/main/guides/scheduling-in-airflow/timetable_manual_run.png)

> This is a simple timetable that could easily be adjusted or built off of to suit other use cases. In general, timetables are completely customizable as long as the methods above are implemented.

Reviewer: ```suggestion
This is a simple timetable that could easily be adjusted to suit other use cases. In general, timetables are completely customizable as long as the methods above are implemented.
```

- **Timetable**: The timetable is a property of a DAG that dictates the data interval and logical date for each DAG Run (i.e. it determines when a DAG will be scheduled).  
- **Run After**: The earliest time the DAG can be scheduled. This date is shown in the Airflow UI, and may be the same as the end of the data interval depending on your DAG's timetable.
- **Backfilling and Catchup**: We won't cover these concepts in depth here, but they can be related to scheduling. We recommend reading [the Apache Airflow documentation](https://airflow.apache.org/docs/apache-airflow/1.10.1/scheduler.html#backfill-and-catchup) on them to understand how they work and whether they're relevant for your use case.

> Note: in this guide we do not cover the `execution_date` concept, which has been deprecated as of Airflow 2.2. If you are using older versions of Airflow, review [this doc](https://airflow.apache.org/docs/apache-airflow/stable/dag-run.html) for more on `execution_date`.

### Parameters

The following parameters are derived from the concepts described above and are important for ensuring your DAG runs at the correct time.

> - **`data_interval_start`**: A datetime object defining the start date and time of the data interval. A DAG's timetable will return this parameter for each DAG Run.

Reviewer: I just realized, if the DAG is returning this does that necessarily make it a parameter? I always assumed parameter is something that a user could control/ configure themselves. Not sure what the solution here is, but I would differentiate this from the other parameters that are configured in code.

Reviewer: I remember changing a lot of instances of this to the parameter name, but looking back maybe that was a misstep. If you change that back to a proper word, you can move these interval terms back up to the "Terms" section.

Author: Yeah these ones are a bit tricky. They're a parameter in the sense that if you're defining a custom timetable, they're something your function has to return (so you are controlling them). But they're also inherent to a DAG Run now, so they'll be created automatically by Airflow in some cases. I do think it makes sense to keep them in the parameter section here, but I'll add some language to clarify.

For the rest of the text, I think it's okay to refer to them either as the parameter or as just plain text. Like "timetable", it is in some ways both a term and a parameter :)

> Note: If your DAG does not need to run on a schedule and will only be triggered manually or externally triggered by another process, you can set `schedule_interval=None`.
    
### Timedelta 
    
If you want to schedule your DAG on a particular cadence (hourly, every 5 minutes, etc.) rather than at a specific time, you can pass a `timedelta` object to the schedule interval. For example, `schedule_interval=timedelta(minutes=30)` will run the DAG every thirty minutes, and `schedule_interval=timedelta(days=1)` will run the DAG every day.
    
> Note: Do not make your DAG's schedule dynamic (e.g. `datetime.now()`)! This will cause an error in the Scheduler.

### Limitations

The scheduling options detailed in the sections above cover many common use cases, but they do have some limitations. For example, it is difficult or impossible to implement situations like:

- Schedule a DAG at different times on different days, like 2pm on Thursdays and 4pm on Saturdays.
- Schedule a DAG daily except for holidays.
- Schedule a DAG at multiple times daily with uneven intervals (e.g. 1pm and 4:30pm).

> In the next section, we'll describe how these limitations were addressed in Airflow 2.2.

Reviewer: ```suggestion
In the next section, we'll describe how these limitations were addressed in Airflow 2.2 with the introduction of timetables.
```
